tests/jwt_test.py

```diff
@@ -49,28 +65,147 @@ def test_refresh_jwt_issuer(network, args):
+    issuer2 = infra.jwt_issuer.JwtIssuer("https://example.issuer2")
+    issuer2.cert_pem = issuer1.cert_pem
+
+    set_issuer_with_keys(network, primary, issuer1, ["kid1"])
+    set_issuer_with_keys(network, primary, issuer2, ["kid1"])
+
+    network.consortium.remove_jwt_issuer(primary, issuer1.name)
+    network.consortium.remove_jwt_issuer(primary, issuer2.name)
+
+
+@reqs.description("Issuer constraint gets overwritten properly for same issuer+kid")
+def test_jwt_same_issuer_constraint_overwritten(network, args):
+    primary, _ = network.find_nodes()
+
+    issuer = infra.jwt_issuer.JwtIssuer("https://example.issuer")
+    keys = issuer.create_jwks_for_kids(["kid1"])
+
+    with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as metadata_fp:
+        json.dump({"issuer": issuer.name}, metadata_fp)
+        metadata_fp.flush()
+        network.consortium.set_jwt_issuer(primary, metadata_fp.name)
+
+    with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as jwks_fp:
+        json.dump(keys, jwks_fp)
+        jwks_fp.flush()
+        network.consortium.set_jwt_public_signing_keys(
+            primary, issuer.name, jwks_fp.name
+        )
+
+    service_keys = get_jwt_keys(args, primary)
+    assert service_keys["kid1"]["issuers"][issuer.name] == issuer.name
+
+    new_constraint = "https://example.issuer/very/specific"
+    keys["keys"][0]["issuer"] = new_constraint
+    with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as jwks_fp:
+        json.dump(keys, jwks_fp)
+        jwks_fp.flush()
+        network.consortium.set_jwt_public_signing_keys(
+            primary, issuer.name, jwks_fp.name
+        )
+
+    service_keys = get_jwt_keys(args, primary)
+    assert service_keys["kid1"]["issuers"][issuer.name] == new_constraint
+
+    network.consortium.remove_jwt_issuer(primary, issuer.name)
+
+
+@reqs.description("Only able to set keys with issuer constraints matching the url")
+def test_jwt_issuer_domain_match(network, args):
+    """Check domains match. Additional subdomains permitted. For example, https://limited.facebook.com
+    may provide keys with issuer constraint https://facebook.com."""
+
+    primary, _ = network.find_nodes()
+
+    issuer = infra.jwt_issuer.JwtIssuer("https://trusted.issuer.com/something")
+    keys = issuer.create_jwks_for_kids(["kid1"])
+
+    with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as metadata_fp:
+        json.dump({"issuer": issuer.name}, metadata_fp)
+        metadata_fp.flush()
+        network.consortium.set_jwt_issuer(primary, metadata_fp.name)
+
+    with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as jwks_fp:
+        json.dump(keys, jwks_fp)
+        jwks_fp.flush()
+        network.consortium.set_jwt_public_signing_keys(
+            primary, issuer.name, jwks_fp.name
+        )
+
+    service_keys = get_jwt_keys(args, primary)
+    assert service_keys["kid1"]["issuers"][issuer.name] == issuer.name
+
+    keys["keys"][0]["issuer"] = "https://issuer.com"
```

Comment:
Can you add a comment or brief `LOG.info` explaining this line? It looks like it's setting a permitted constraint (in contrast to the later `garbage` constraints), but its not obvious to me why we're setting this nested field directly.

Reply:
I'm not sure what explanation to add around this. We set it because the `keys` are retrieved above as a parsed JSON, so in the test we overwrite this field only (which is the issuer constraint) to make sure only certain constrains can be set.

Doesn't look sophisticated to me, but I don't find it ugly also.